test/integration/singlecluster/controller/failurerecovery/pod_termination_controller_test.go

```diff
@@ -0,0 +1,104 @@
+	reachableNodeName          = "reachable-node"
+	forcefulTerminationTimeout = 2 * time.Second
+)
+
+func createTerminatingPod(p *corev1.Pod) {
+	ginkgo.GinkgoHelper()
+
+	util.MustCreate(ctx, k8sClient, p)
+	gomega.ExpectWithOffset(1, k8sClient.Delete(ctx, p)).To(gomega.Succeed())
+}
+
+var _ = ginkgo.Describe("Pod termination controller", ginkgo.Ordered, ginkgo.ContinueOnFailure, func() {
+	var ns *corev1.Namespace
+	var matchingPodWrapper *testingpod.PodWrapper
+
+	ginkgo.BeforeAll(func() {
+		ns = util.CreateNamespaceFromPrefixWithLog(ctx, k8sClient, "pod-fr-namespace-")
+		nodes := []corev1.Node{
+			*testingnode.MakeNode(unreachableNodeName).
+				NotReady().
+				Taints(corev1.Taint{Key: corev1.TaintNodeUnreachable, Effect: corev1.TaintEffectNoSchedule}).
+				Obj(),
+			*testingnode.MakeNode(reachableNodeName).
+				Ready().
+				Obj(),
+		}
+		util.CreateNodesWithStatus(ctx, k8sClient, nodes)
+
+		matchingPodWrapper = testingpod.MakePod("matching-pod", ns.Name).
+			StatusPhase(corev1.PodPending).
+			TerminationGracePeriod(1).
+			NodeName(unreachableNodeName).
+			Annotation(constants.SafeToForcefullyTerminateAnnotationKey, constants.SafeToForcefullyTerminateAnnotationValue)
+	})
+
+	ginkgo.It("forcefully terminates pods that opt-in, scheduled on unreachable nodes", func() {
+		matchingPod := matchingPodWrapper.Clone().Obj()
+		createTerminatingPod(matchingPod)
+
+		gomega.Eventually(func(g gomega.Gomega) {
+			g.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: matchingPod.Name, Namespace: matchingPod.Namespace}, matchingPod)).
+				To(gomega.Succeed())
+			g.Expect(matchingPod.Status.Phase).Should(gomega.Equal(corev1.PodFailed))
+		}, forcefulTerminationTimeout, util.Interval).Should(gomega.Succeed())
+	})
+
+	ginkgo.It("does not forcefully terminate matching pods that did not opt-in or are running on healthy nodes", func() {
+		nonMatchingPod := matchingPodWrapper.
+			Clone().
+			Name("non-matching-pod").
+			Annotation(constants.SafeToForcefullyTerminateAnnotationKey, constants.SafeToForcefullyTerminateAnnotationValue).
+			Obj()
+		createTerminatingPod(nonMatchingPod)
+
+		podOnHealthyNode := matchingPodWrapper.Clone().Name("healthy-pod").NodeName(reachableNodeName).Obj()
+		createTerminatingPod(podOnHealthyNode)
+
+		gomega.Consistently(func(g gomega.Gomega) {
+			// Non-matching pod is left untouched.
+			g.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: nonMatchingPod.Name, Namespace: nonMatchingPod.Namespace}, nonMatchingPod)).
+				To(gomega.Succeed())
+			g.Expect(nonMatchingPod.Status.Phase).Should(gomega.Equal(corev1.PodPending))
+
+			// Healthy pod is left untouched.
+			g.Expect(k8sClient.Get(ctx, types.NamespacedName{Name: podOnHealthyNode.Name, Namespace: podOnHealthyNode.Namespace}, podOnHealthyNode)).
+				To(gomega.Succeed())
+			g.Expect(podOnHealthyNode.Status.Phase).Should(gomega.Equal(corev1.PodPending))
+		}, forcefulTerminationTimeout, util.Interval).Should(gomega.Succeed())
```

Comment:
I think this could flake potentially, if the test code starts a check just before forcefulTerminationTimeout elapses, but loses CPU for a while. Then the controller can make progress and transition to Failed, resulting in the test failing. 

TBH I think we need two tests with different configuration:
- (1.) forcefulTerminationTimeout=2s, and just wait to see we transition to Failed eventually
- (2.) forcefulTerminationTimeout=10s, and verify that we stay in Pending for 1s or 2s

Reply:
That's why I made the check timeout 2 seconds instead of lower:
1. We have to use 1 second for the deletion grace period.
2. I set the forceful termination grace period to 1 millisecond.

So technically, there are `2s - 1s - 1ms = 999ms` for the forceful termination to occur and be seen in the test. Since the check interval is 250ms, there are going to be 2/3 checks "after" the termination should have occurred.